pkg/tcl/testworkflowstcl/testworkflowcontroller/controller.go

```diff
@@ -0,0 +1,357 @@
+			}
+			if v.Value.Reason == "SuccessfulCreate" {
+				result.QueuedAt = v.Value.CreationTimestamp.Time
+			}
+			if v.Value.Type == "Normal" {
+				continue
+			}
+			w.SendValue(Notification{
+				Timestamp: v.Value.CreationTimestamp.Time,
+				Log:       fmt.Sprintf("%s (%s) %s\n", v.Value.CreationTimestamp.Time.Format(time.RFC3339Nano), v.Value.Reason, v.Value.Message),
+			})
+		}
+
+		// Emit the result with queue time
+		if result.QueuedAt.IsZero() {
+			w.SendError(errors.New("job is in unknown state"))
+			return
+		}
+		w.SendValue(Notification{Result: result.Clone()})
+
+		// Wait for the pod initialization
+		for v := range WatchPodPreEvents(ctx, c.podEvents, 0).Stream(ctx).Channel() {
+			if v.Error != nil {
+				w.SendError(v.Error)
+				continue
+			}
+			if v.Value.Reason == "Scheduled" {
+				result.StartedAt = v.Value.CreationTimestamp.Time
+				result.Status = common.Ptr(testkube.RUNNING_TestWorkflowStatus)
+			}
+			if v.Value.Type == "Normal" {
+				continue
+			}
+			w.SendValue(Notification{
+				Timestamp: v.Value.CreationTimestamp.Time,
+				Log:       fmt.Sprintf("%s (%s) %s\n", v.Value.CreationTimestamp.Time.Format(time.RFC3339Nano), v.Value.Reason, v.Value.Message),
+			})
+		}
+
+		// Emit the result with start time
+		if result.StartedAt.IsZero() {
+			w.SendError(errors.New("pod is in unknown state"))
+			return
+		}
+		w.SendValue(Notification{Result: result.Clone()})
+
+		// Wait for the initialization container
+		for v := range WatchContainerPreEvents(ctx, c.podEvents, "tktw-init", 0).Stream(ctx).Channel() {
+			if v.Error != nil {
+				w.SendError(v.Error)
+				continue
+			}
+			if v.Value.Reason == "Created" {
+				result.Initialization.QueuedAt = v.Value.CreationTimestamp.Time
+			} else if v.Value.Reason == "Started" {
+				result.Initialization.StartedAt = v.Value.CreationTimestamp.Time
+				result.Initialization.Status = common.Ptr(testkube.RUNNING_TestWorkflowStepStatus)
+			}
+			if v.Value.Type == "Normal" {
+				continue
+			}
+			w.SendValue(Notification{
+				Timestamp: v.Value.CreationTimestamp.Time,
+				Log:       fmt.Sprintf("%s (%s) %s\n", v.Value.CreationTimestamp.Time.Format(time.RFC3339Nano), v.Value.Reason, v.Value.Message),
+			})
+		}
+
+		// Emit the result with start time
+		if result.Initialization.StartedAt.IsZero() {
+			w.SendError(errors.New("init container is in unknown state"))
+			return
+		}
+		w.SendValue(Notification{Result: result.Clone()})
+
+		// Watch the initialization container logs
+		pod := (<-c.pod.Any(ctx)).Value
+		for v := range WatchContainerLogs(ctx, c.clientSet, c.podEvents, c.namespace, pod.Name, "tktw-init").Stream(ctx).Channel() {
+			if v.Error != nil {
+				w.SendError(v.Error)
+				continue
+			}
+			// TODO: Calibrate clock with v.Value.Hint or just first/last timestamp here
+			w.SendValue(Notification{
+				Timestamp: v.Value.Time,
+				Log:       fmt.Sprintf("%s %s\n", v.Value.Time.Format(time.RFC3339Nano), string(v.Value.Log)),
+			})
+		}
```

Comment:
should we break it on small methods?

Reply:
Probably yes, but it will take a significant amount of time - splitting the operations would likely require creating few `struct`s to share the state across these operations, and will cause more complicated phases (to cleanup contexts and close channels properly across distributed functions).

Do you think that it will be enough if I'll add a follow-up ticket for that, or do you think that it needs to be done in this phase?

Reply:
add the ticket, it should be enough for now

Reply:
Thanks, created https://linear.app/kubeshop/issue/TKC-1713/split-testworkflow-watcher-to-smaller-functions 👍